README
.---------------------------------------------------------.
| ECU README - last revised Wed Sep  2 13:52:15 EDT 1998  |
`---------------------------------------------------------'

ECU is a asynchronous serial and telnet communications program
for these environments:

  SCO XENIX System V/286          Support has been dropped; ECU
                                  is now too large. Try versions
                                  prior to 3.30.

  SCO XENIX System V/386          ECU is stable on SCO XENIX/386
                                  (telnet support attempted
                                  if your DS has inet support)

  SCO UNIX System V/386           ECU is very robust on SCO UNIX,
  SCO ODT 1.x,2.0,3.0             ODT and OS5 (telnet support works
  SCO Open Server 5               if your DS has inet support)
  SCO UnixWare 2.1

  SunOS 4.1.x                     a robust, stable, limited subset
                                  (telnet support automatic)

  ISC 386/ix 2.2 or later         Ports to these systems are
  ISC System V Release 4          not supported as regularly
  ESIX System V Release 4         and I cannot vouch for
                                  them at time of release
                                  PLEASE GIVE ME FEEDBACK!

  Linux 1.1.58                    tested well enough under RedHat 5
    or later                      (telnet support automatic)

  HP-UX 9.01                      reasonably functional; a few
                                  rough edges with keyboards and
                                  modem control signals

  Motorola Delta [68]8K SVR40     stable; some testing completed
                                  (telnet support automatic)

  Motorola Delta [68]8K SVR32     compiles OK with Green Hills and GCC;
                                  (now known to NOT work)
                                  no machine to debug upon

  NetBSD                          Still under development;
                                  ANSI emulation disabled until
                                  work on termcap support is complete

  Solaris 2.x                     compiles; some testing done; needs more

  FreeBSD                         very reliable and full-featured port

  AIX                             Motorola's version anyway; compiles OK
                                  but untested

-----------------------------------------------------------------------
ECU 4.0 is released in the month of ECU's eleventh anniversary!!!!
ECU 4.12 Thu Dec 18 06:11:05 EST 1997
ECU 4.30 Wed Sep  2 13:52:48 EDT 1998
-----------------------------------------------------------------------

ECU (Extended Call Utility) is a research and engineering
communications program originally written for users of SCO UNIX
V.3.2/386 and XENIX V on 80286 and 80386 systems.  Support for
other systems has been added and further porting is possible with
"minor" effort to other systems based on or similar to UNIX
System V (see README.porting).  

ECU provides the classic terminal communications facility of
passing keyboard data to a serial line and incoming data to the
computer video display.  In addition, a dialing directory, a
function key mapping feature, session logging, and other
basic features are available.

ECU presents to the host a flexible "ANSI" terminal type,
accepting any valid video control sequences from MS-DOS or SCO
documentation as of late 1990.  It also fares well, though
imperfectly, with Sun and VT-100 video control sequences.
Standards are great: everybody should have one, especially if
they call it "ANSI". For more information, refer to the manual
section titled "ANSI Filter."  (NOTE: the porter to BSD did
not engineer the ANSI terminal emulation. Therefore the incoming
data is passed directly to your native screen. This may be
repaired in a future release.)  As of ECU 3.37, the ANSI filter
may be turned on and off.

ECU supports numerous file transfer protocols: as of this
writing, XMODEM, XMODEM/CRC, XMODEM-1K, YMODEM/CRC Batch,
ZMODEM/CRC-16, ZMODEM/CRC-32, and C-Kermit 5 are
supported.  For more information, refer to the manual sections
describing the individual interactive and procedure file transfer
commands. (Usage of serial file transfer protocols over telnet/inet
is not supported.)

A very flexible procedure (script) language is also incorporated
to automate many communications tasks.  In addition to augmenting
interactive tasks, by using shell scripts and ECU procedures, ECU
can perform batch-style communications sessions in an entirely
"unattended" fashion.

For applications too unwieldy for the procedure language,
"ecufriend" programs are supported.  Friends are spawned by ECU
having access to the shared memory segment containing an
ECU-managed "screen image" and other data and having use of the
attached communications line.

The doc subdirectory has all of the .txt files used to produce
ecu.man, the manual of sorts for the program.  A copy of it is
reluctantly included (net.bandwidth) for those who do not have
my squirrely 1985 Xenix/286 nroff.  (Gotta rework that doc someday.)

*Please* take the time to read the (tedious) manuals and READMEs.
This will do me honor and yourself justice because there are a
lot of goodies in here, many of which are not traditional
features you'll be looking for.

-------------------------------------------------------------------------

RIGHTS TO USE

This program, it sources, objects and utilities are in the public
domain.

Alas, placing "popular" programs in the public domain can cause
peculiar problems.  As others adopt, modify and redistribute the
code, it tends to fall apart.  I regularly get mail from folks
who have found fragmented or damaged buckets of code with my
e-mail address in it.  Invariably, it has come from some off-net
BBS where someone has tried to PKZIP it or something :).

Your license to redistribute this version of ECU is, of course,
unlimited.  Even though I'll spend hours poring over patches
trying to help a user, my patience with folks who modify my code
without CLEARLY marking it so is at end.  

These are my requests:

1.  If you use or redistribute this program, it should retain the
ECU name unless you ask me.  This means the base name of the
program executable must be "ecu" and the program must announce
itself with the strings "ECU" and "wht@n4hgf".

2.  If you modify the program "substantially" you should not
redistribute it without asking me.  For purposes of this
paragraph, "substantially" refers to changes of more than a
bug-fix or cosmetic nature.  For larger changes, read
README.porting and, if appropriate, submit patches to
wht@wht.net.  If they "pass muster" (we are usually
easy), then the changes will be incorporated into a released
patch level of ECU.

You can give modified versions to a associate, but if s/he gives
it to someone else, does the fourth party know not to contact me
for bug fixes?

3.  You are still free to use substantial code fragments or
entire logical subsections of the program in any way you see fit,
obeying of course any other copyrights and requests made by other
authors whose code fragments appear within ECU.

4.  Anybody who has contributed to ECU development will find it
awfully easy to get my permission to do anything they want with
ECU ("Noooooooooooooooooo ...  not THAT!")

--------------------------------------------------------------------

ACKNOWLEDGMENTS

MANY THANKS to those who helped me improve the program, especially
upaya!tbetz, ache@hq.demos.su (now ache@astral.msk.su), spel@hippo.ru.ac.za,
bel@trout.nosc.mil, dhmadsen@icaen.uiowa.edu, dug@kd4nc.atl.ga.us,
jts@ki4xo, jsm@n4vu.atl.ga.us, lamy@glsys.in-berlin.de, cma@tridom,
tabbs!aris, neal@clkwrka, extel@quagga.ru.ac.za,
mjb@mjbtn, tmcsys.uucp!lothar, mju@mudos.ann-arbor.mi.us
elastic!fche, genrad!rob and spooley@compulink.co.uk.  There were
lots of others and I know I've forgotten someone who helped a
great deal; I apologize.

Very special thanks go to Dion Johnson and Bob Lewis at SCO for
their untiring and generous support.

Lothar Hirschbiegel <aega84!lh> did the ISC SVR4 port -- THANKS,
Lothar!

Joseph H Buehler <jhpb@sarto.budd-lake.nj.us> extended the SVR4
port to ESIX -- THANKS, Joesph!

Bob Lewis <robertle@sco.com> proofread the manual (as of 3.30 --
don't blame them for errors creeping in since then <smile>).
This is tedious work and special thanks go to them.

Robert Lipe has contributed greatly to ECU in various esoteric areas
(wild signals and tty driver dependencies particularly).  Also,
I do not believe there ever has been an alpha revision Robert has
not compiled and tested at least a bit (usually thoroughly).
Robert also did early stabs at the ports to Solaris 2.x.
Thanks, Robert!

Robert "Bob" Broughton ported ECU to LINUX with the help of
Toomas Losin <a776@mindlink.bc.ca>.  THANKS, Bob and Toomas!

Carl Wuebker ported ECU to HP-UX.  THANKS, Carl!

Daniel Harris ported ECU to NetBSD and helped settle POSIX
termios/SYSV termio issues.  THANKS, Daniel!

Andrey Chernov <ache@nagual.ru> (formerly <ache@astral.msk.su>)
ported ECU to FreeBSD.  His port was not only flawless and
seamless, but exquisite.  All I had to do was patch -p0 and make.
He was patient with the minor fractures I made in pursuing my own
agenda in parallel.  (Parallel source development across
timezones is difficult.  Corollary: Linux is a miracle.)

Don Yuniskis <dgy@rtd.com> also contributed to solidifying
FreeBSD support.

Robert Lipe, Bob Friesenhahn and Roger Fujii all
submitted input and patches for Solaris 2.x .

Thanks go to Pat Davitt for checking the guy out under
XENIX/386 2.3.1.

-------------------------------------------------------------------------

CHEST THUMPING

There is a reason for all the edit notes and funky comment
formats in ECU.  As a bonus for ECU's 11th anniversary, docxtr.c
and edlists.c should amuse True Hackers(TM) at least momentarily.
These tools have often 1) amused me when I wanted to burn up some
time on a new CPU, 2) unconfused me when I had to look at ancient
code, and 3) shown me how much of my life I've thrown away at the
keyboard.

cd ecu-src-dir/
cc -o docxtr docxtr.c

CHANGES
Changes for 3.30 since 3.24:

1.  An ANSI CUP (cursor position) with x or y values exceeding
screen size should set the cursor to the bottom or right of the
screen, not the top or left.  ECU did the latter; it is fixed.
The DSR (device status report) handler was broken also.  It is
fixed, and differentiation is now made between a type 5 and 6 DSR
request.  As a result of all this, the X11 resize program will work
with ECU.

2. The command line -l handler now treats an argument
with a leading equals sign as a Devices type rather than
an explicit line.  That is,

    ecu -l=VSAT

will search for an ``idle'' line matching any entry
in your HDB Devices file whose first field is "VSAT".
The term ``idle'' is strictly correct only on SCO machines
or where the utmp/ecuungetty code has been made to work.
The baud rate is automatically set to the highest speed
specified by the Devices entry, overriding any value set
by the -b command line switch.

Note that specifying a logical (directory entry name)
rather than an actual (numeric) phone number will cause the
directory entry choice of line to override the line specified on
by -l.  This has always been true of -lttyxx treatment as well.

Perhaps one day, I'll have time to make a directory entry
specifying "Any" *not* override -l and initiate it's own search for a
free line (based as it is on baud rate and type == "ACU" alone).
There is no time now.

3. The command line usage has been updated to show what
command line variables are overriden by a logical phone number.
Again, no time to put it in the manual.

4. The integer operators << and >> were added.

5. The procedure command eeod (erase to end of display) was added.

6.  Previously, you could bind a function key to a character
sequence to be transmited over the line.  You may now bind a
function key to a procedure invocation.  If the first token in
the third field of a key definition is "proc", the rest of the
third field is taken to be a procedure invocation.

Consider the following example entry in ~/.ecu/keys:

test
    F1:VT100 F1:esc O q
    F2:doit:proc doit arg1 arg2 argn
                 ^  64 chars max   ^
                 `-----------------'

The maximum length of the procedure name, arguments and
whitespace must be less than 64 characters; any excess is
silently discarded.

Procedure doit might, for example, send a prompt to a remote to
begin an XMODEM transaction, then start an XMODEM receive to
get the file.  (This helps make up for the unsuitability of XMODEM's
C or NAK header for automatic download startup.)

7. The fkey definition file (~/.ecu/keys in the documentation)
can be located in the library directory (usually /usr/local/lib/ecu).
This has always been in the code.  Now it is documented.

8. The code for saving a phone directory lacked good error detection.
This has been fixed.

9.  -D__STDC__=0 was added to gcc flags.

10.  The default for SCO UNIX configuration is now 'v' not 'u',
making 3.2v4/ODT 2.0 the default.

11. A discussion of how to "finish" a remote C-kermit server was
added to README.Q-and-A.

12. Under SunOS, DTR was not properly controlled.  This has been fixed.

13. Configuration for use with gcc now requires you have gcc 1.40
or later.  If you say you have gcc 2.3.3 or later, you will get -O6
compilations (full global optimaztion).  If you encounter problems
with the software compile with -O6, reconfigure answering no to
the 2.3.3 question and remake.  Please let me know of any problems.

14. -Wuninitialized was added to gcc compilations.
This is only used if you answered yes to the 2.3.3 question

15. ECU works again on XENIX 286 thanks to vancleef@netcom.com
(Henry van Cleef)

    a. CFLAGS from "-M2let128" to "-M2e -i"
    b. ZMODEM code unsigned Baudrate to unsigned long

(Not a bad port given I haven't seen a 286 in 5 years!)

16.  An experimental USR 2400 dialer (dceUSR24v.c/dialgUSR24v)
has been added to gendial for attended voice line autodialer
applications.  It is written for speed over reliability, so so
the careful initialization found in other gendial dialers is
absent.  This means dialgUSR24v works best with a modem that is
dedicated to the voice autodial function.

As delivered, dialgUSR24v will prefix a '1' to a number passed to it if
the number is ten digits long and has [01] as its second digit.
(I wanted to word click with xterm and my phone directory is an
ancient flat file with no '1' prefix to be found.)
You may need to adopt, hack and improve.

DialgUSR24v, like all gendial dialers, are supplied "as-is" in
hope the code may be useful. (Okay, and also to promote the
excellent SCO solution to modem madness.)

17. Previously, a Devices-specified  SCO dialer program had to
have a full pathname specified for ECU to use it (otherwise,
ECU would treat the token as the name of an entry in Dialers).
Now, when ECU sees a simple name in a Devices entry, it checks
for an executable in the UUCP library directory before searching
Dialers.

18. PgUp and PgDn have been added to the directory menu repertorie.

PgUp   ^U    page up

ecu.c
/*+-----------------------------------------------------------------------
	ecu.c - Extended Calling Unit/Call Utility/Call UNIX/whatever
	wht@wht.net

  Defined functions:
	main(argc, argv)
	xmtr()

   "Now tell me, old man, what's your name?"
   "Holloway, sir. Charles William Holloway."
   "Oh yes, the town's librarian."
   "I have the honor, sir, and have had for many years."
   "I believe all that time spent living through other men's lives,
other men's dreams, what a waste."
   "Sometimes, a man can learn more from other men's dreams than he
can from his own.  Come visit me sir, if you wish to improve your
education and I may improve yours."  -- Something Wicked This Way Comes

------------------------------------------------------------------------*/
/*+:EDITS:*/
/*:04-26-2000-11:15-wht@bob-RELEASE 4.42 */
/*:12-19-1997-21:45-wht@kepler-add -k */
/*:12-18-1997-05:06-wht@sidonia-AIX port: sbrk and getpgrp */
/*:05-10-1997-13:53-wht@gyro-strncpy not strcpy to shm->ttyname */
/*:01-24-1997-02:36-wht@yuriatin-SOURCE RELEASE 4.00 */
/*:10-24-1996-03:11-wht@yuriatin-T switch now incremental sw for proc_trace */
/*:09-11-1996-19:59-wht@yuriatin-3.48-major telnet,curses,structural overhaul */
/*:08-21-1996-16:57-wht@fep-add -C switch for show_config */
/*:08-20-1996-12:39-wht@kepler-locale/ctype fixes from ache@nagual.ru */
/*:12-12-1995-14:22-wht@kepler-notify [no line attached] on keystroke */
/*:12-06-1995-13:30-wht@n4hgf-termecu w/errno -1 consideration */
/*:12-03-1995-19:57-wht@gyro-use Setuid */
/*:11-23-1995-11:20-wht@kepler-source control 3.37 for tsx-11 */
/*:11-14-1995-10:23-wht@kepler-3.37.80-source control point: SOCKETS */
/*:11-13-1995-12:07-wht@kepler-xmtr() can live with no line open */
/*:11-12-1995-01:11-wht@gyro-add switch -n to turn off rcvr_ansi_filter */
/*:11-12-1995-00:18-wht@gyro-show_telnet_traffic now in shm */
/*:11-10-1995-18:32-wht@gyro-add switch -z for show_telnet_traffic */
/*:11-04-1995-21:03-root@wwtp1-move windows rattle for quirk earlier in run */
/*:10-14-1995-16:17-wht@kepler-call build_valid_baud_string */
/*:09-17-1995-16:28-wht@kepler-remove obsolete #if 0 code */
/*:06-14-1995-19:08-wht@n4hgf-do not use setuid if root is real user */
/*:06-12-1995-15:03-wht@n4hgf-if ecu has uucp euid, make use of it */
/*:06-12-1995-14:13-wht@n4hgf-no more root euid nice code */
/*:01-12-1995-15:19-wht@n4hgf-apply Andrew Chernov 8-bit clean+FreeBSD patch */
/*:05-04-1994-04:38-wht@n4hgf-ECU release 3.30 */
/*:08-16-1993-17:19-wht@n4hgf-aid share debug with report_initial_line() */
/*:08-07-1993-20:38-wht@n4hgf-add xmtr_wfp_debug_hack */
/*:12-20-1992-12:08-wht@n4hgf-continue -l=Devices support work */
/*:12-04-1992-20:57-wht@n4hgf-add -l=Devices support */
/*:09-10-1992-13:58-wht@n4hgf-ECU release 3.20 */
/*:08-22-1992-15:38-wht@n4hgf-ECU release 3.20 BETA */
/*:08-16-1992-03:43-wht@n4hgf-add -F funckeytype */
/*:08-16-1992-03:08-wht@n4hgf-head off another POSIX plot */
/*:04-30-1992-00:49-wht@n4hgf-remove obsolete -c */
/*:04-24-1992-21:59-wht@n4hgf-more SCO tty name normalizing */
/*:04-17-1992-18:19-wht@n4hgf-"default" keyset read if present */
/*:02-16-1992-02:39-wht@n4hgf-add -P phonedir switch for rll@sco */
/*:02-16-1992-01:41-wht@n4hgf-turn off xterm_title */
/*:08-11-1991-19:56-wht@n4hgf-soup up -l for ISC vs. SCO */
/*:08-06-1991-13:02-wht@n4hgf-jpm@logixwi fix: HZ getenv test wrong sense */
/*:07-29-1991-17:57-wht@n4hgf-add memstat */
/*:07-25-1991-12:55-wht@n4hgf-ECU release 3.10 */
/*:07-17-1991-07:04-wht@n4hgf-avoid SCO UNIX nap bug */
/*:04-27-1991-01:52-wht@n4hgf-overhaul revision numbers */
/*:03-17-1991-13:44-wht@n4hgf-nice and uid revision */
/*:01-09-1991-22:31-wht@n4hgf-ISC port */
/*:11-30-1990-19:04-wht@n4hgf-new ttyinit parameter - see TTYINIT_... */
/*:11-28-1990-15:58-wht@n4hgf-add non-ansi terminal support */
/*:08-14-1990-20:39-wht@n4hgf-ecu3.00-flush old edit history */

#ifdef __FreeBSD__
#include <locale.h>
#endif
#include "ecu.h"
#include "dvent.h"
#include "esd.h"
#include "procedure.h"
#include "ecukey.h"
#include "ecuxkey.h"
#include "termecu.h"
#include "ecu_pwd.h"

char *getenv();
void xmtr_SIGINT_handler();

extern char *makedate;		 /* temporary make date */
extern char *revstr;		 /* ecunumrev.c */
extern char *revision_modifier;	/* ecunumrev.c */
extern ESD *icmd_prompt;
extern char kbdeof;			 /* current input EOF */
extern char kbdeol2;		 /* current secondary input EOL */
extern char kbdeol;			 /* current input EOL */
extern char kbderase;		 /* current input ERASE */
extern char kbdintr;		 /* current input INTR */
extern char kbdkill;		 /* current input KILL */
extern char kbdquit;		 /* current input QUIT */
extern UINT tcap_LINES;
extern UINT tcap_COLS;
extern UINT32 colors_current;
extern int there_is_hdb_on_this_machine;
extern char phonedir_name[PHONEDIR_NAME_SIZE];

char *eculibdir = CFG_EcuLibDir;
char *dash_f_funckeytype;
char hello_str[128];		 /* msg printed upon BOJ */
char errmsg[128];
char initial_procedure[128] = "";
char *default_tty = CFG_DefaultTty;	/* configurable default tty name */
int init_proc_argc = 0;
char *init_proc_argv[MAX_PARGV];
CFG_PidType xmtr_pid = 0;
int quit_on_init_proc_fail = 0;
int quit_on_init_proc_done = 0;
int rc_ep_has_run = 0;
struct TIMEB starting_timeb;

extern int hertz;					 /* HZ from environ or sys/param.h */
extern UINT32 hzmsec;				 /* clock period in msec rounded up */

#ifdef CFG_Malloc3X
char *startbrk;
char *startsp;

#endif

char *_rc = "_rc";			 /* _rc.ep */

/*+-----------------------------------------------------------------------
	xmtr() --  copy stdin to comm line

  THE INITIAL PROCESS EXECUTES THIS PROCESS UNTIL PROGRAM TERMINATION

------------------------------------------------------------------------*/
void
xmtr()
{
	UINT xmtr_char;
	char nlchar = NL;
	char tmpc;

	xmtr_wfp_debug_hack();	 /* don't ask */
	ttymode(1);
	xmtr_signals();
	while (1)
	{

		xmtr_char = ttygetc(1);

		if (xmtr_char & 0x100)
		{
			kbd_escape(xmtr_char);
			continue;
		}

		if (shm->Liofd <= 0)
		{
			UINT32 colors_at_entry = colors_current;

			setcolor(colors_notify);
			fputs("[no line attached]", se);
			setcolor(colors_at_entry);
			ff(se, "\r\n");
			continue;
		}

		lputc(xmtr_char);
		if (!shm->Lfull_duplex)
		{					 /* echo character if asked */
			tmpc = xmtr_char;
			write(TTYERR, &tmpc, 1);
		}

		if (xmtr_char == CRET)
		{
#if defined(CFG_TelnetOption)
			if (shm->Ltelnet && !shm->Ltelnet_raw)
				lputc(0);
#endif
			if (shm->Ladd_nl_outgoing)
				lputc('\n');
			if (!shm->Lfull_duplex)
				write(TTYERR, &nlchar, 1);
		}
	}
	/* NOTREACHED */

}							 /* end of xmtr */

/*+-------------------------------------------------------------------------
	keyboard_test_only() - diagnose keyboard problems
--------------------------------------------------------------------------*/
void
keyboard_test_only()
{
	ttyinit(TTYUSE_NORMAL);
	if (tty_not_char_special)
	{
		ff(se, "Cannot test keyboard unless it is stanmdard input\n");
		exit(1);
	}
	ttymode(1);				 /* put user console in `raw' mode */
	kbd_test();				 /* in ecutty.c */
	ttymode(0);				 /* restore keyboard */
	termecu(0);

}							 /* end of keyboard_test_only */

/*+-------------------------------------------------------------------------
	main(argc,argv)

  main() program forks to create rcvr process; then main()
  becomes the xmtr process
------------------------------------------------------------------------*/
int
main(argc, argv)
int argc;
char **argv;
{
	int swchar;
	int itmp;
	UINT32 colors_save;
	struct passwd *pw;
	char *get_ttyname();
	extern char *optarg;
	extern int optind;

#ifdef CFG_Malloc3X

	/*
	 * absolutely must be first on the list
	 */
	startbrk = sbrk(0);		 /* initial break */
	startsp = (char *)&swchar;	/* initial sp */
	(void)mallopt(M_MXFAST, 256);
	(void)mallopt(M_NLBLKS, 64);
	(void)mallopt(M_GRAIN, 64);
#endif

#ifdef __FreeBSD__
	setlocale(LC_ALL, "");
#endif

	setbuf(stderr, NULL);	 /* rarely necessary */

	build_valid_baud_string();

	Ftime(&starting_timeb);	 /* get startup time */
	xmtr_signals();			 /* catch xmtr signals */
	xmtr_pid = (CFG_PidType) getpid();

	/*
	 * get this off quick, cause we'll be busy for a little while on 286
	 */
	build_revision_string();
	sprintf(hello_str, "ecu %s", revstr);
#ifdef M_I286
	ff(se, "%s\n", hello_str);
#endif

	/*
	 * initialize 'hertz'/'hzmsec' and  Nap() implimentation
	 */
	init_Nap();

	/*
	 * init uucp euid processing
	 */
	uid = getuid();
	euid = geteuid();
	if (pw = getpwnam("uucp"))
		uid_uucp = pw->pw_uid;
	endpwent();

	/*
	 * if we are not root, nor uucp (unlikely) and we are euid uucp, then
	 * use Setuid() to access uucp owned lines
	 */
	setuid_uucp = (uid) && (uid_uucp) && (uid != euid) && (euid == uid_uucp);
	if ((uid != euid) && Setuid(uid))
	{
		ff(se, "\r\n");
		perror("Setuid(uid) failed");
		ff(se, "\r\n");
		exit(-1);
	}

	/*
	 * Initialize shared memory segment. This must be done before any uses
	 * are made of the Lermio variable see ecushm.c and ecu.h.
	 */
	shm_init();

	/*
	 * ecu has its own keyboard handler ... blessing and curse since it
	 * depends on little outside help but can be tedious to configure for
	 * a new host or even instance
	 */
	keyset_init();			 /* intialize keyset */
	keyset_read("default");

	/*
	 * make ~/.ecu if necessary ... but chmod 700 the directory even if it
	 * already existed (needs to be very early before lots of init)
	 */
	make_ecu_subdir();

	get_curr_dir(curr_dir, sizeof(curr_dir));
	cd_array_init();		 /* read %cd directory list */

	hdb_init();
	var_init();				 /* initialize procedure variables */
	poutput_init();			 /* initialize procedure output */

	icmd_prompt = esdalloc(64);
	set_default_escape_prompt();

	/*
	 * init line variables
	 */
	memset(shm->Lline, 0, sizeof(shm->Lline));
	shm->Liofd = -1;		 /* no line open now */
	shm->Lbitrate = CFG_DefaultBitRate;	/* from config.c run */
	shm->Lparity = CFG_DefaultParity;	/* from config.c run */
	if (shm->Lparity == 'n')
		shm->Lparity = 0;
	shm->Ltelno[0] = 0;		 /* no telephone number for remote yet */
	shm->Llogical[0] = 0;	 /* no logical name for remote yet */
	shm->Lrname[0] = 0;		 /* no logical name for remote yet */
	shm->Ldescr[0] = 0;		 /* no description for remote yet */
	shm->Lconnected = 0;	 /* not connected */
	shm->Ladd_nl_incoming = 0;	/* dont add nl to incoming cr */
	shm->Ladd_nl_outgoing = 0;	/* dont add nl to outgoing cr */
	shm->Lfull_duplex = 1;	 /* assume full duplex */
	shm->Lmodem_already_init = 0;	/* modem has not been initialized */
	shm->Lxonxoff = IXON | IXOFF;	/* default to xon/xoff protocol */
	shm->xmtr_pid = (CFG_PidType) getpid();
	shm->xmtr_ppid = (CFG_PidType) getppid();
#if defined(CFG_GetpgrpVoidArg)
	shm->xmtr_pgrp = (CFG_PidType) getpgrp();
#else
	shm->xmtr_pgrp = (CFG_PidType) getpgrp(0);
#endif
	strncpy(shm->tty_name, get_ttyname(), sizeof(shm->tty_name));
	shm->tty_name[sizeof(shm->tty_name) - 1] = 0;
	shm->shm_revision = SHM_REV;
	shm->ttyuse = TTYUSE_NORMAL;
	shm->terminating = 0;

	while ((swchar = getopt(argc, argv, "kCDF:HNP:Tb:defhl:no:p:tz")) != -1)
	{
		switch (swchar)
		{
			case 'k':
				keyboard_test_only();
				/* NOTREACHED */
			case 'b':
				if (valid_baud(shm->Lbitrate = atoi(optarg)) == -1)
				{
					ff(se, "invalid bit rate %u\n", shm->Lbitrate);
					usage();
				}
				break;
			case 'l':
				shm->Lline[0] = 0;
				if ((*optarg != '=') && (*optarg != '/'))
					strcpy(shm->Lline, "/dev/");
				strncat(shm->Lline, optarg,
					sizeof(shm->Lline) - strlen(shm->Lline) -
					strlen(optarg) - 1);
				shm->Lline[sizeof(shm->Lline) - 1] = 0;
				break;
			case 'p':
				strncpy(initial_procedure, optarg, sizeof(initial_procedure));
				initial_procedure[sizeof(initial_procedure) - 1] = 0;
				break;
			case 'h':
				shm->Lfull_duplex = 0;
				break;
			case 'f':
				shm->Lfull_duplex = 1;
				break;
			case 'd':
				quit_on_init_proc_fail = 1;
				break;
			case 'D':
				quit_on_init_proc_done = 1;
				break;
			case 't':
				shm->Ladd_nl_incoming = 1;
				shm->Ladd_nl_outgoing = 1;
				break;
			case 'e':
				shm->Lparity = 'e';
				break;
			case 'o':
				shm->Lparity = 'o';
				break;
			case 'N':
				shm->ttyuse = TTYUSE_FORCE_SIMPLE;
				break;
			case 'T':
				proc_trace++;
				break;
#if defined(CFG_TelnetOption)
			case 'R':
				shm->Ltelnet_raw = 1;
				break;
#endif
			case 'P':
				strncpy(phonedir_name, optarg, PHONEDIR_NAME_SIZE);
				phonedir_name[PHONEDIR_NAME_SIZE - 1] = 0;
				break;
			case 'C':
				show_config();
				termecu(0);
				break;
			case 'F':
				dash_f_funckeytype = optarg;
				break;
			case 'n':
				shm->rcvr_ansi_filter = 0;
				break;
			case 'z':
				shm->show_telnet_traffic = 1;
				break;
			case '?':
				usage();
		}
	}

	/*
	 * check a few options for validity
	 */
	if (!initial_procedure[0] &&
		(quit_on_init_proc_done || quit_on_init_proc_fail))
	{
		ff(se, "no -D/-d without -p\n");
		usage();
	}

	ttyinit(shm->ttyuse);	 /* init console tty mode handler */
	ttymode(3);				 /* put user console in `raw' mode but SIGINT
							  * terms prog */

	if (tty_not_char_special)
		quit_on_init_proc_done = 1;
	else
	{

		/*
		 * rattle curses once - fixes quirk/bug I can't find
		 */
		windows_start();
		windows_end(0);
		fflush(so);
		tcap_clear_screen();
	}

#if defined(WHT2) || defined(XTERM_FRIEND)

	/*
	 * if xterm, put notice in title bar but this really should be done in
	 * _rc.ep
	 */
	xterm_title("ECU", 0);
#endif

	/*
	 * do the _rc.ep execution
	 */
	if (find_procedure(_rc))
	{
		if (do_proc(1, &_rc))
		{
			if (quit_on_init_proc_fail || quit_on_init_proc_done)
				termecu(TERMECU_INIT_PROC_ERROR);
		}
	}
	rc_ep_has_run = 1;

	/*
	 * check out line
	 */
	if (!shm->Lline[0])
	{
		if (!there_is_hdb_on_this_machine)
			strcpy(shm->Lline, default_tty);
		else
		{
			DVE *tdve;

			if (tdve = hdb_choose_Any(shm->Lbitrate))
				sprintf(shm->Lline, "/dev/%s", tdve->line);
			else
				strcpy(shm->Lline, default_tty);
		}
	}
	else if (shm->Lline[0] == '=')	/* Devices lookup */
	{
		DVE *tdve = 0;
		char acutype[128];

		if (!there_is_hdb_on_this_machine)
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fputs("[cannot use -l=Devices ... HDB not available]", se);
			tcap_stand_end();
			ff(se, "\r\n");
			usage();
		}
		strncpy(acutype, shm->Lline + 1, sizeof(acutype));
		acutype[sizeof(acutype) - 1] = 0;
		if (!(tdve = getdvtype(shm->Lline + 1)))
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fprintf(se, "[no Devices line of type '%s']", acutype);
			tcap_stand_end();
			ff(se, "\r\n");
			errno = -1;
			termecu(TERMECU_SVC_NOT_AVAIL);
		}
		shm->Lline[0] = 0;
		shm->Lbitrate = tdve->high_baud;
		setdvent();			 /* rewind for assured passthru/enddvent by
							  * hdb_choose */
		if (!(tdve = hdb_choose_Device(acutype, shm->Lbitrate)))
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fprintf(se, "[no idle Devices line of type '%s' at %u baud]",
				acutype, shm->Lbitrate);
			tcap_stand_end();
			ff(se, "\r\n");
			errno = -1;
			termecu(TERMECU_SVC_NOT_AVAIL);
		}
		strcpy(shm->Lline, "/dev/");
		strncat(shm->Lline, tdve->line, sizeof(shm->Lline) - strlen(shm->Lline));
		shm->Lline[sizeof(shm->Lline) - 1] = 0;
	}

	report_initial_line();	 /* in #ifdef'ed debug event logging context */

	/*
	 * either present startup screen or run initial procedure or both
	 */
	if (initial_procedure[0])
	{
		init_proc_argv[0] = initial_procedure;
		init_proc_argc = 1;
		for (itmp = optind; itmp < argc; itmp++)
		{
			if (*argv[itmp] != '-')
			{
				if (init_proc_argc == MAX_PARGV)
				{
					ff(se, "too many arguments to initial procedure\r\n");
					errno = -1;
					termecu(TERMECU_USAGE);
				}
				init_proc_argv[init_proc_argc++] = argv[itmp];
			}
		}

		if (do_proc(init_proc_argc, init_proc_argv))
		{
			if (quit_on_init_proc_fail || quit_on_init_proc_done)
				termecu(TERMECU_INIT_PROC_ERROR);
		}
		proc_file_reset();
		colors_save = colors_current;
		setcolor(colors_notify);
		pputs("[procedure finished]");
		setcolor(colors_save);
		pputs("\n");
		if (quit_on_init_proc_done)
		{
			errno = -1;
			termecu(TERMECU_OK);
		}
		if (shm->Liofd < 0)
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fputs("[no line attached by initial procedure]", se);
			tcap_stand_end();
			ff(se, "\r\n");
			if (quit_on_init_proc_fail)
				termecu(TERMECU_INIT_PROC_ERROR);
			tcap_stand_out();
			fputs("[press ESC to exit or SPACE for setup menu]", se);
			tcap_stand_end();
			itmp = ttygetc(0);
			ff(se, "\r\n");
			if (itmp == ESC)
				termecu(TERMECU_OK);
			shm->Llogical[0] = 0;
			shm->Ltelno[0] = 0;
			shm->Ldescr[0] = 0;
			setup_screen((char *)0);
		}
		else
			start_rcvr_process(0);
	}
	else
		/* no initial procedure */
	{
		if (!tty_not_char_special)
			setup_screen((optind < argc) ? argv[optind] : (char *)0);
		else
		{
			errno = -1;
			pprintf("Cannot run ecu from non-tty without initial procedure.\n");
			termecu(TERMECU_UNRECOVERABLE);
		}
	}

	/* enter xmtr operation */
	xmtr();
	termecu(TERMECU_OK);
	return (0);				 /* never get here, but keep gcc optim from
							  * complaining */
	/* NOTREACHED */
}							 /* end of main */

/* end of ecu.c */
/* vi: set tabstop=4 shiftwidth=4: */
